### PR TITLE
Extending `fork` to return the ThreadId.

### PR DESCRIPTION
**Pull request by 365andreas**
Trying to implement logging in Sim and integrate it with `iohk-monitoring-framework` package we encountered the need to acquire the `ThreadId` of a thread. Consequently the MonadFork class was extended with a `ThreadId` type, `myThreadId` function and `fork` changed to return the `ThreadId`.

```
class (Monad m, Show (ThreadId m)) => MonadFork m where
  type ThreadId m :: *

  myThreadId :: m (ThreadId m)
  fork       :: m () -> m (ThreadId m)
```

Relevant comment on the PR set up for the aforementioned goal (logging in Sim): https://github.com/input-output-hk/ouroboros-network/pull/298#discussion_r257505533

### CODE REVIEW

**Comment by 365andreas** [comment]
> LGTM; Could you add a test for Sim monad.

@coot Yes
What should this test do?
Spawn several threads and check if they have different `ThraedId`s and especially each new one has the  Id of the previous incremented by 1?

**Comment by 365andreas** [comment]
Like this:
https://github.com/input-output-hk/ouroboros-network/blob/75d1db8e2904a3d1fe7920568cf0f96217e92a39/io-sim/test/Test/IOSim.hs#L224-L247 ?

**Comment by coot** [comment]
Indeed.

**Comment by coot** [comment]
Now you don't need to use probes since `runSim` can return values from `SimM` (and soon probes will be removed).

**Comment by dcoutts** [comment]
@365andreas Sorry, ended up doing the same (copying your suggestion) in #304.

Would you mind rebasing on that? The class changes may be the same but the extra tests are always appreciated.

**Comment by 365andreas** [comment]
> @365andreas Sorry, ended up doing the same (copying your suggestion) in #304.
> 
> Would you mind rebasing on that? The class changes may be the same but the extra tests are always appreciated.

Ok.

**Comment by 365andreas** [comment]
rebased on `dcoutts/sim-async-exceptions` (#304)

File: io-sim/src/Control/Monad/IOSim.hs [hunk 1]
```diff
@@ -95,6 +95,7 @@ data SimA s a where
   Catch        :: Exception e =>
                   SimA s a -> (e -> SimA s a) -> (a -> SimA s b) -> SimA s b
 
+  GetTid       :: (ThreadId -> SimA s b) -> SimA s b
```
Comment (dcoutts) [inline io-sim/src/Control/Monad/IOSim.hs]: This is here already as `GetThreadId` below.

File: ouroboros-consensus/src/Ouroboros/Consensus/Util/STM.hs [hunk 2]
```diff
@@ -17,6 +17,7 @@ module Ouroboros.Consensus.Util.STM (
   , simChaChaT
   ) where
 
+import           Control.Monad (void)
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Consensus/Util/STM.hs]: Needed?

File: io-sim/test/Test/IOSim.hs [hunk 3]
```diff
@@ -26,7 +26,7 @@ import           Control.Monad.Class.MonadSTM
 import           Control.Monad.Class.MonadThrow
 import           Control.Monad.Class.MonadTimer
 import           Control.Monad.Class.MonadSay
-import           Control.Monad.IOSim
+import           Control.Monad.IOSim hiding (ThreadId)
```
Comment (dcoutts) [inline io-sim/test/Test/IOSim.hs]: This is no longer exported.

File: io-sim/src/Control/Monad/IOSim.hs (hunk 1 above)
Comment (365andreas) [reply io-sim/src/Control/Monad/IOSim.hs]: deleted

File: ouroboros-consensus/src/Ouroboros/Consensus/Util/STM.hs (hunk 2 above)
Comment (365andreas) [reply ouroboros-consensus/src/Ouroboros/Consensus/Util/STM.hs]: No. deleted

File: io-sim/test/Test/IOSim.hs (hunk 3 above)
Comment (365andreas) [reply io-sim/test/Test/IOSim.hs]: fixed

**Comment by 365andreas** [comment]
@dcoutts it seems ready

**Comment by coot** [comment]
Thanks @365andreas; LGTM

**Comment by 365andreas** [comment]
@dcoutts should I rebase on `dcoutts/sim-async-exceptions` ?

File: cabal.project [hunk 4]
```diff
@@ -17,5 +17,4 @@ package ouroboros-consensus
 source-repository-package
   type: git
   location: https://github.com/input-output-hk/iohk-monitoring-framework
-  tag: 09b269cd1926303902e8dea3c8703b6e588004f5
-
+  tag: e5336c722da520272717fd6b17c25c46bd72ebb9
```
Comment (dcoutts) [inline cabal.project]: Is this the right one now? See https://github.com/input-output-hk/ouroboros-network/commit/aaf377b1256b46d2117f727934a7920ccd8ad58f

Comment (365andreas) [reply cabal.project]: Yes will remove mine.

**Pull request merged**